src/Components/Tooltip.re

```diff
@@ -0,0 +1,109 @@
+open Oni_Core;
+open Revery;
+open Revery.UI;
+
+module Constants = {
+  let delay = Time.ms(400);
+  let offsetX = 10;
+  let offsetY = 10;
+};
+
+// MODEL
+
+type t = {
+  text: string,
+  x: float,
+  y: float,
+};
+
+// TOOLTIP
+
+module Tooltip = {
+  module Styles = {
+    open Style;
+
+    let tooltip = (~theme: Theme.t, ~x, ~y) => [
+      position(`Absolute),
+      left(int_of_float(x) + Constants.offsetX),
+      top(int_of_float(y) + Constants.offsetY),
+      backgroundColor(theme.menuBackground),
+      paddingVertical(3),
+      paddingHorizontal(8),
+      boxShadow(
+        ~xOffset=3.,
+        ~yOffset=3.,
+        ~blurRadius=5.,
+        ~spreadRadius=0.,
+        ~color=Color.rgba(0., 0., 0., 0.2),
+      ),
+    ];
+
+    let tooltipText = (~theme: Theme.t, ~font: UiFont.t) => [
+      fontFamily(font.fontFile),
+      fontSize(font.fontSize),
+      color(theme.menuForeground),
+      textWrap(TextWrapping.NoWrap),
+    ];
+  };
+
+  let make = (~text, ~x, ~y, ~theme, ~font, ()) =>
+    <View style={Styles.tooltip(~theme, ~x, ~y)}>
+      <Text style={Styles.tooltipText(~theme, ~font)} text />
+    </View>;
+};
+
+// OVERLAY
+
+module Overlay: {
+  let setTooltip: t => unit;
+  let clearTooltip: unit => unit;
+
+  let make:
+    (~key: React.Key.t=?, ~theme: Theme.t, ~font: UiFont.t, unit) => element;
+} = {
+  let internalSetTooltip = ref(_ => ());
+  let debouncedSetTooltip = {
+    let clearTimeout = ref(() => ());
+
+    maybeTooltip => {
+      clearTimeout^();
+      switch (maybeTooltip) {
+      | Some(tooltip) =>
+        clearTimeout :=
+          Tick.timeout(
+            () => internalSetTooltip^(_ => Some(tooltip)),
+            Constants.delay,
+          )
+      | None => internalSetTooltip^(_ => None)
+      };
+    };
+  };
+
+  let setTooltip = tooltip => debouncedSetTooltip(Some(tooltip));
+  let clearTooltip = () => debouncedSetTooltip(None);
+
+  let%component make = (~theme, ~font, ()) => {
+    let%hook (current, setCurrent) = Hooks.state(None);
+    internalSetTooltip := setCurrent;
```

Comment:
I've used a simpler mechanism here to communicate between the trigger/hotspot/anchor and the overlay components while still triggering a re-render.

In the context menu I used the state to pass information between them and trigger re-render. Here I use a hooks instead, and then assign the hook's setter to a global ref, which can then be called by the trigger component, through some convenience functions to cover over the rough edges. Much more contained, and works pretty well I think.

Reply:
Neat!